pygsti/models/model.py

```diff
@@ -1210,8 +1210,126 @@ def split_circuit(self, circuit, erroron=('prep', 'povm'), split_prep=True, spli
             povm_lbl = None
 
         return prep_lbl, circuit, povm_lbl
+    
+    def split_circuits(self, circuits, erroron=('prep', 'povm'), split_prep=True, split_povm=True):
+        """
+        Splits a circuit into prep_layer + op_layers + povm_layer components.
 
-    def complete_circuit(self, circuit):
+        If `circuit` does not contain a prep label or a
+        povm label a default label is returned if one exists.
+
+        Parameters
+        ----------
+        circuit : list of Circuit
+            A list of circuits, possibly beginning with a state preparation
+            label and ending with a povm label.
+
+        erroron : tuple of {'prep','povm'}
+            A ValueError is raised if a preparation or povm label cannot be
+            resolved when 'prep' or 'povm' is included in 'erroron'.  Otherwise
+            `None` is returned in place of unresolvable labels.  An exception
+            is when this model has no preps or povms, in which case `None`
+            is always returned and errors are never raised, since in this
+            case one usually doesn't expect to use the Model to compute
+            probabilities (e.g. in germ selection).
+
+        split_prep : bool, optional
+            Whether to split off the state prep and return it as `prep_label`.  If
+            `False`, then the returned preparation label is always `None`, and is
+            not removed from `ops_only_circuit`.
+
+        split_povm : bool, optional
+            Whether to split off the POVM and return it as `povm_label`.  If
+            `False`, then the returned POVM label is always `None`, and is
+            not removed from `ops_only_circuit`.
+
+        Returns
+        -------
+        list of tuples containing 
+        prep_label : Label or None
+        ops_only_circuit : Circuit
+        povm_label : Label or None
+        """
+
+        #get the tuple of povm labels to avoid having to access through dict
+        #many times.
+        primitive_prep_labels_tup = self.primitive_prep_labels
+        primitive_povm_labels_tup = self.primitive_povm_labels
+        primitive_prep_labels_set = set(primitive_prep_labels_tup)
+        primitive_povm_labels_set = set(primitive_povm_labels_tup)
+
+        #precompute unique default povm labels.
+        unique_sslbls = set([ckt._line_labels for ckt in circuits])
+        default_povm_labels = {sslbls:self._default_primitive_povm_layer_lbl(sslbls) for sslbls in unique_sslbls}
+        
+        if split_prep and split_povm: #can avoid some duplicated effort in this case.
+            split_circuits = []
+            for ckt in circuits:
+                if len(ckt) > 0 and ckt[0] in primitive_prep_labels_set:
```

Comment:
Stylistically I prefer avoiding nested/local function definitions where possible and I am not sure it would improve readability over using the sets directly. Your comment did make it occur to me that it would be (very slightly) more efficient to only construct the sets we need given the value of `split_prep` and `split_povm`, and doing so does improve code locality a bit (at the expense of a bit more repetition, but sometimes these two desiderata are at odds).